extensions/eclipse-che-theia-prefs-provider/src/browser/prefs-provider.ts

```diff
@@ -0,0 +1,130 @@
+import { WorkspaceService } from '@theia/workspace/lib/browser';
+import WorkspaceClient, { IRemoteAPI } from '@eclipse-che/workspace-client';
+import { che } from '@eclipse-che/api';
+
+export interface WorkspaceWithDevfile extends che.workspace.Workspace {
+    devfile: che.workspace.devfile.Devfile;
+}
+
+@injectable()
+export class PrefsProvider implements FrontendApplicationContribution {
+
+    constructor(
+        @inject(EnvVariablesServer) private readonly envVariablesServer: EnvVariablesServer,
+        @inject(PreferenceServiceImpl) private readonly preferenceService: PreferenceServiceImpl,
+        @inject(WorkspaceService) private readonly workspaceService: WorkspaceService,
+    ) { }
+
+    private getPluginsProperties(workspace: che.workspace.Workspace): [string, string][] {
+        if (!!workspace.devfile) {
+            return this.getPropsFromDevfile(workspace);
+        } else if (!!workspace.config) {
+            return this.getPropsFromConfig(workspace);
+        }
+        throw new TypeError('Can\'t get either "config" or "devfile" of current workspace configuration.');
+    }
+
+    private getPropsFromConfig(workspace: che.workspace.Workspace): [string, string][] {
+        const attributes = workspace.config!.attributes;
+        if (!attributes) {
+            return [];
+        }
+
+        return Object.keys(attributes)
+            .filter((attrKey: string) => attrKey.indexOf('plugin.') === 0 && attrKey.indexOf('.preference.') !== -1)
+            .map((attrKey: string) => <[string, string]>[attrKey.split('.preference.')[1], attributes[attrKey]]);
+    }
+
+    private getPropsFromDevfile(workspace: che.workspace.Workspace): [string, string][] {
+        const components = workspace.devfile!.components;
+        if (!components) {
+            throw new TypeError('Can\'t get "components" of current workspace "devfile" section.');
+        }
+
+        return components
+            // tslint:disable-next-line:no-any
+            .filter((component: che.workspace.devfile.Component) => (<any>component).preferences)
+            // tslint:disable-next-line:no-any
+            .map((component: che.workspace.devfile.Component) => (<any>component).preferences)
+            .reduce((result: [string, string][], preferences: { [key: string]: string }) => {
+                Object.keys(preferences).forEach(key => {
+                    result.push(<[string, string]>[key, preferences[key]]);
+                });
+                return result;
+            }, []);
+    }
+
+    private async setPluginProperties(props: [string, string][]): Promise<void> {
+        await this.workspaceService.roots;
+        const workspace = this.workspaceService.workspace;
+        if (!workspace) {
+            throw new Error('Failed to get Theia workspace.');
+        }
+        for (const [key, value] of props) {
+            await this.preferenceService.set(key, value, PreferenceScope.Workspace, workspace.uri);
```

Comment:
Shouldn't the scope here be "default"? Otherwise, how can the user override this preference?

Reply:
Good catch!
Actually, default scope preference provider is not designed to set preferences ([preference-contribution.ts#L255](https://github.com/theia-ide/theia/blob/master/packages/core/src/browser/preferences/preference-contribution.ts#L255)). So "user" scope is the only option here, isn't it?

Reply:
Maybe an approach based on PreferenceProvider would work? I would take the discussion to the Theia chat, tbh.

Reply:
I'm not sure about overriding preferences in default scope, so let's consider other options first.

Setting recommended preferences in user scope is not an option here because user scope preferences are synchronizing between Che workspaces.

With workspace scope, we can make the extension working in the following way:
- set preferences if workspace scope doesn't contain them
- otherwise, ask user to override preferences with recommended ones.

What do you think?

Reply:
> set preferences if workspace scope doesn't contain them

+1

Reply:
I still think the right thing to do would be to try an approach on PreferenceProvider. I'm sure Anton Kosyakov could provide feedback if that makes sense. Not an expert here.

Reply:
I'm merging the PR as is, and [here](https://github.com/eclipse/che-theia/issues/243) is the issue to continue the discussion